src/flow/src/compute/render/reduce.rs

```diff
@@ -0,0 +1,1057 @@
+
+    // no future updates should exist here
+    if arrange.read().get_next_update_time(&now).is_some() {
+        err_collector.push_err(
+            InternalSnafu {
+                reason: "No future updates should exist in the reduce distinct arrangement",
+            }
+            .build(),
+        );
+    }
+
+    send.give(ret);
+}
+
+/// eval accumulable reduce plan by eval aggregate function and reduce the result
+///
+/// TODO(discord9): eval distinct by adding distinct input arrangement
+///
+/// invariant: it'is assumed `kv`'s time is always <= now,
+/// since it's from a Collection Bundle, where future inserts are stored in arrange
+///
+/// the data being send is just new rows that represent the new output after given input is processed
+///
+/// i.e: for example before new updates comes in, the output of query `SELECT sum(number), count(number) FROM table`
+/// is (10,2(), and after new updates comes in, the output is (15,3), then the new row being send is ((15, 3), now, 1)
+///
+/// while it will also update key -> accums's value, for example if it is empty before, it will become something like
+/// |offset| accum for sum | accum for count |
+/// where offset is a single value holding the end offset of each accumulator
+/// and the rest is the actual accumulator values which could be multiple values
+fn reduce_accum_subgraph(
+    arrange: &ArrangeHandler,
+    distinct_input: &Option<Vec<ArrangeHandler>>,
+    kv: impl IntoIterator<Item = KeyValDiffRow>,
+    accum_plan: &AccumulablePlan,
+    SubgraphArg {
+        now,
+        err_collector,
+        scheduler,
+        send,
+    }: SubgraphArg,
+) {
+    let AccumulablePlan {
+        full_aggrs,
+        simple_aggrs,
+        distinct_aggrs,
+    } = accum_plan;
+    let mut key_to_vals = BTreeMap::<Row, Vec<(Row, repr::Diff)>>::new();
+
+    for ((key, val), _tick, diff) in kv {
+        // it is assumed that value is in order of insertion
+        let vals = key_to_vals.entry(key).or_default();
+        vals.push((val, diff));
+    }
+
+    let mut all_updates = Vec::with_capacity(key_to_vals.len());
+    let mut all_outputs = Vec::with_capacity(key_to_vals.len());
+
+    // lock the arrange for write for the rest of function body
+    // so to prevent wide race condition since we are going to update the arrangement by write after read
+    // TODO(discord9): consider key-based lock
+    let mut arrange = arrange.write();
+    for (key, value_diffs) in key_to_vals {
+        let col_diffs = {
+            let row_len = value_diffs[0].0.len();
+            // split value_diffs into columns
+            (0..row_len)
+                .map(|i| {
+                    value_diffs
+                        .iter()
+                        .map(|(row, diff)| (row.get(i).cloned().unwrap(), *diff))
+                        .collect_vec()
+                })
+                .collect_vec()
+        };
```

Comment:
I think we can extract this section into a function and then make some necessary checks (ensure equal number of columns in all rows)